concordia/views.py

```diff
@@ -209,25 +211,58 @@ def registration_rate(self, group, request):
 @method_decorator(never_cache, name="dispatch")
 class ConcordiaRegistrationView(RatelimitMixin, RegistrationView):
     form_class = UserRegistrationForm
+    ratelimit_group = "registration"
     ratelimit_key = "ip"
     ratelimit_rate = registration_rate
     ratelimit_method = "POST"
-    ratelimit_block = True
+    ratelimit_block = settings.RATELIMIT_BLOCK
 
 
 @method_decorator(never_cache, name="dispatch")
 class ConcordiaLoginView(RatelimitMixin, LoginView):
-    ratelimit_key = "ip"
+    ratelimit_group = "login"
+    ratelimit_key = "post:username"
     ratelimit_rate = "3/15m"
     ratelimit_method = "POST"
-    ratelimit_block = True
+    ratelimit_block = False
+
+    def post(self, request, *args, **kwargs):
+        blocked = is_ratelimited(
+            request,
+            group=self.ratelimit_group,
+            key=self.ratelimit_key,
+            method=self.ratelimit_method,
+            rate=self.ratelimit_rate,
+        )
+        recent_captcha = (
+            time.time() - request.session.get("captcha_validation_time", 0)
+        ) < 86400
+
+        if blocked and not recent_captcha:
+            form = self.get_form()
+            form.fields["captcha"] = CaptchaField()
+            return self.form_invalid(form)
+        else:
+            return super().post(request, *args, **kwargs)
 
 
 def ratelimit_view(request, exception=None):
-    template_name = "429.html"
     status_code = 429
-    template = loader.get_template(template_name)
-    return HttpResponse(template.render(), status=status_code)
+
+    if "json" in request.META["HTTP_ACCEPT"].lower():
+        response = JsonResponse({"exception": str(exception), "status": status_code})
+    else:
+        template_name = "429.html"
+        template = loader.get_template(template_name)
+
+        response = HttpResponse(
+            template.render({"exception": exception}), status=status_code
+        )
+
+    response["Retry-After"] = 15 * 60
```

Comment:
Does the ratelimiter pass information on the request about the triggered rate limit period? It'd be nice if `Retry-After` was based on the actual value, although there's not too much harm from a few extra retries.

Reply:
It doesn't as far as I can tell. You actually need to know the rate being used in order to even call the helper function to get a True/False of whether the request was ratelimited.

Reply:
Looking at the code confirms that it only sets the boolean `limited` value. I'm inclined not to spend more time exposing this now.